serial_test/src/code_lock.rs

```diff
@@ -12,16 +12,18 @@ lazy_static! {
 fn check_new_key(name: &str) {
     // Check if a new key is needed. Just need a read lock, which can be done in sync with everyone else
     let new_key = {
-        let unlock = LOCKS.read().unwrap();
+        let unlock = LOCKS.read();
         !unlock.deref().contains_key(name)
     };
     if new_key {
         // This is the rare path, which avoids the multi-writer situation mostly
-        LOCKS
-            .write()
-            .unwrap()
-            .deref_mut()
-            .insert(name.to_string(), ReentrantMutex::new(()));
+        let lock = LOCKS.upgradable_read();
+        if lock.deref().contains_key(name) {
+            return; // another writer got here between unlock of the read and re-acquire
+        }
+        let mut upgraded_lock =
+            RwLockUpgradableReadGuard::<'_, HashMap<String, ReentrantMutex<()>>>::upgrade(lock);
+        upgraded_lock.insert(name.to_string(), ReentrantMutex::new(()));
```

Comment:
Isn't this the same race condition once again? Imagine that both threads hold the `UpgradableReadGuard` and both check that the key is not contained at the same time. Then both race for the `upgrade` to a `WriteGuard` and the winner inserts the first Mutex, the looser replaces it with a different one.

Reply:
That shouldn't happen (see https://docs.rs/lock_api/0.4.5/lock_api/struct.RwLock.html#method.upgradable_read "The calling thread will be blocked until there are no more writers or other upgradable reads which hold the lock") but I'm seeing bits of test locks as well which I'm still sifting through.

Reply:
You're right, this should work because no two upgradable readers can hold the lock at the same time.